scanner/scanner.go

```diff
@@ -0,0 +1,361 @@
+		log.Fatalf("Either source is not a blob or destination is not a local folder")
+	}
+	srcLevel, err := cmd.DetermineLocationLevel(cca.Source.Value, cca.FromTo.From(), true)
+
+	if err != nil {
+		return nil, err
+	}
+
+	dstLevel, err := cmd.DetermineLocationLevel(cca.Destination.Value, cca.FromTo.To(), false)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a Remote resource resolver
+	// Giving it nothing to work with as new names will be added as we traverse.
+	var containerResolver cmd.BucketToContainerNameResolver
+	containerResolver = cmd.NewS3BucketNameToAzureResourcesResolver(nil)
+	if cca.FromTo == common.EFromTo.GCPBlob() {
+		containerResolver = cmd.NewGCPBucketNameToAzureResourcesResolver(nil)
+	}
+//	existingContainers := make(map[string]bool)
+//	var logDstContainerCreateFailureOnce sync.Once
+	seenFailedContainers := make(map[string]bool) // Create map of already failed container conversions so we don't log a million items just for one container.
+
+	dstContainerName := ""
+
+	filters := cca.InitModularFilters()
+
+	processor := func(object cmd.StoredObject) error {
+		// Start by resolving the name and creating the container
+		if object.ContainerName != "" {
+			// set up the destination container name.
+			cName := dstContainerName
+			// if a destination container name is not specified OR copying service to container/folder, append the src container name.
+			if cName == "" || (srcLevel == cmd.ELocationLevel.Service() && dstLevel > cmd.ELocationLevel.Service()) {
+				cName, err = containerResolver.ResolveName(object.ContainerName)
+
+				if err != nil {
+					if _, ok := seenFailedContainers[object.ContainerName]; !ok {
+						cmd.WarnStdoutAndScanningLog(fmt.Sprintf("failed to add transfers from container %s as it has an invalid name. Please manually transfer from this container to one with a valid name.", object.ContainerName))
+						seenFailedContainers[object.ContainerName] = true
+					}
+					return nil
+				}
+
+				object.DstContainerName = cName
+			}
+		}
+
+		// If above the service level, we already know the container name and don't need to supply it to makeEscapedRelativePath
+		if srcLevel != cmd.ELocationLevel.Service() {
+			object.ContainerName = ""
+
+			// When copying directly TO a container or object from a container, don't drop under a sub directory
+			if dstLevel >= cmd.ELocationLevel.Container() {
+				object.DstContainerName = ""
+			}
+		}
+
+		srcRelPath := cca.MakeEscapedRelativePath(true, isDestDir, object)
+		//dstRelPath := cca.MakeEscapedRelativePath(false, isDestDir, object)
+                message := fmt.Sprintf("ScannerRun: %v", srcRelPath)
+                glcm.Info(message)
+                return nil
+	}
+	finalizer := func() error {
+		return nil
+	}
+
+	return cmd.NewCopyEnumerator(traverser, filters, processor, finalizer), nil
```

Comment:
why not use the existing initEnumerator function?

Reply:
The reason for not using initEnumerator function is we want Processing function and Finalizer Function should work according to our requirement. That diverge lead me to use initResourceEnumerator and create our own initEnumerator function.